### ISSUE TITLE
Optimizing TSD and thread cache layout

### ISSUE THREAD

**Issue by interwq** [jemalloc/jemalloc#701]
Fast path tcache performance is critical as it's expected to be accessed very frequently. We aim to improve the fast path significantly with the following two changes:

1) Embed tcache into TSD:
Currently each tsd stores a pointer to the automatic tcache:
https://github.com/jemalloc/jemalloc/blob/dev/include/jemalloc/internal/tsd_structs.h#L19
This adds a memory dereference on the fast path, and potentially a cache miss. We can avoid this by embedding the tcache struct into tsd directly. Since we do not allow switching auto tcache, this should work fine w/o breaking APIs.

2) Optimize tcache and tsd for a more compact layout. We'd like have higher density of the tcache bins to improve cache locality. Currently tbins are 32 bytes, which will be cut down to 24. Also pack TSD tighter.

**Comment by interwq** [comment]
@jasone : to embed tcache into tsd, my current plan is to split the tbins into two bins: small and large, basically:

tcache_t {
    ....
    tbin_t small_bins[NBINS];
    tbin_t *large_bins; /*Dynamically sized*/
}
So that we embed small bins for better locality, while keep the ability to set lg_tcache_max, i.e. nhbins. Does that make sense to you? Or you think we should just get rid of lg_tcache_max and make the size static?

**Comment by jasone** [comment]
I think we should make ```large_bins``` have ```NSIZES - NBINS``` elements so that no dynamic sizing is necessary, and probably place tcache at the end of tsd so that the unused elements don't break up the utilized cachelines.

**Comment by interwq** [comment]
That sounds reasonable. For the unused large bins (binind > nhbins), probably not allocating their avail arrays?

**Comment by jasone** [comment]
Ah right, definitely avoid allocating more avail slots than actually used.

**Comment by interwq** [comment]
The first half of this (embedding tcache into tsd) is implemented in #736. Tested with @davidtgoldblatt using test/stress/microbench, which shows a ~15% improvement on malloc / free / etc. Likely this made the fast path more pipeline friendly. A sample output:
With #736:
```
        100000000 iterations, malloc=2189009us, mallocx=2275009us, ratio=1:0.962
        test_malloc_vs_mallocx: pass
        100000000 iterations, free=2192009us, dallocx=2223009us, ratio=1:0.986
        test_free_vs_dallocx: pass
        100000000 iterations, dallocx=2190009us, sdallocx=2326009us, ratio=1:0.942
        test_dallocx_vs_sdallocx: pass
        100000000 iterations, malloc_usable_size=2679011us, sallocx=2683011us, ratio=1:0.999
        test_mus_vs_sallocx: pass
        100000000 iterations, sallocx=2710011us, nallocx=2636011us, ratio=1:1.028
        test_sallocx_vs_nallocx: pass
        --- pass: 5/5, skip: 0/5, fail: 0/5 ---
```
Before (https://github.com/jemalloc/jemalloc/commit/d4e98bc0b27b2ed660002e1603fa353d5c89eb91):
```
        100000000 iterations, malloc=2540010us, mallocx=2704011us, ratio=1:0.939
        test_malloc_vs_mallocx: pass
        100000000 iterations, free=2568010us, dallocx=2546010us, ratio=1:1.009
        test_free_vs_dallocx: pass
        100000000 iterations, dallocx=2528010us, sdallocx=2486010us, ratio=1:1.017
        test_dallocx_vs_sdallocx: pass
        100000000 iterations, malloc_usable_size=3018012us, sallocx=3235013us, ratio=1:0.933
        test_mus_vs_sallocx: pass
        100000000 iterations, sallocx=3254013us, nallocx=3004012us, ratio=1:1.083
        test_sallocx_vs_nallocx: pass
```
I'll start working on splitting tcache into tcache_fast and tcache slow, to be placed at the beginning and end of tsd respectively. Also make tbin layout more compact.

**Comment by interwq** [comment]
Closed via #736 and #752.
